iceoryx2-services/discovery/src/service_discovery/service.rs

```diff
@@ -0,0 +1,331 @@
+                publish_subscribe
+                    .publisher_builder()
+                    .initial_max_slice_len(1024)
+                    .allocation_strategy(AllocationStrategy::PowerOfTwo)
+                    .create()
+                    .map_err(|e| {match e {
+                        iceoryx2::port::publisher::PublisherCreateError::ExceedsMaxSupportedPublishers => CreationError::PublisherAlreadyExists,
+                        iceoryx2::port::publisher::PublisherCreateError::UnableToCreateDataSegment => CreationError::PublisherCreationError,
+                    }})?
+            );
+        }
+
+        let mut notifier = None;
+        if discovery_config.send_notifications {
+            let event = node
+                .service_builder(service_name())
+                .event()
+                .max_notifiers(1)
+                .max_listeners(discovery_config.max_listeners)
+                .create()
+                .map_err(|e| {
+                    match e {
+                        iceoryx2::service::builder::event::EventCreateError::IsBeingCreatedByAnotherInstance => CreationError::ServiceAlreadyExists,
+                        iceoryx2::service::builder::event::EventCreateError::AlreadyExists => CreationError::ServiceAlreadyExists,
+                        _ => CreationError::ServiceCreationFailure,
+                    }
+                })?;
+
+            let port = event.notifier_builder().create().map_err(|e| match e {
+                iceoryx2::port::notifier::NotifierCreateError::ExceedsMaxSupportedNotifiers => {
+                    CreationError::NotifierAlreadyExists
+                }
+            })?;
+            notifier = Some(port);
+        }
+
+        let mut tracker = Tracker::<S>::new();
+
+        if discovery_config.sync_on_initialization {
+            tracker.sync(iceoryx_config);
+        }
+
+        Ok(Service::<S> {
+            discovery_config: discovery_config.clone(),
+            iceoryx_config: iceoryx_config.clone(),
+            _node: node,
+            publisher,
+            notifier,
+            tracker,
+        })
+    }
+
+    /// Processes service changes and emits events/notifications.
+    ///
+    /// This function should be called periodically to detect changes in the service
+    /// landscape and emit appropriate events and notifications.
+    ///
+    /// # Returns
+    ///
+    /// A result containing a tuple of:
+    /// - A vector of references to services that were added since the last call
+    /// - A vector of services that were removed since the last call
+    ///
+    /// # Errors
+    ///
+    /// Returns a `SpinError` if there was an error publishing events or sending
+    /// notifications.
+    #[allow(clippy::type_complexity)] // Using a type alias for return value is less clear in this
+                                      // case IMO
+    pub fn spin(&mut self) -> Result<(Vec<&ServiceDetails<S>>, Vec<ServiceDetails<S>>), SpinError> {
```

Comment:
Here we force heap on the user. Could we change the API so that the user provides two callables. One is called for every added service and one is called for every removed service and the get a `&ServiceDetails<S>` as input argument.

With this approach we do not force heap on the user. It is perfectly `no_std` and the user has still the freedom to push everything into a heap vector if they like to.

So the signature would be:
```rust
pub fn spin<FAddedService: FnMut(&ServiceDetails<S>), FRemovedService: FnMut(&ServiceDetails<S>)>(&mut self, mut f_added_service: FAddedService, mut f_removed_service: FRemovedService) -> Result<(), SpinError> {...
```